> I would have thought the casts would need to include 'const' to avoid assignments

No, code like:

  (type)(lvalue) = somevalue;

is ill-formed.  Old versions of gcc allowed it as an extension.
```
tony@venus:.../perl/git$ gcc -c castlvalue.c
castlvalue.c: In function ‘f’:
castlvalue.c:5:14: error: lvalue required as left operand of assignment
    5 |     (int *)x = y;
      |              ^
castlvalue.c: In function ‘g’:
castlvalue.c:9:12: error: lvalue required as left operand of assignment
    9 |     (int)z = a;
      |            ^
tony@venus:.../perl/git$ clang-14 -c castlvalue.c
castlvalue.c:5:5: error: assignment to cast is illegal, lvalue casts are not supported
    (int *)x = y;
    ^~~~~~~~ ~
castlvalue.c:9:5: error: assignment to cast is illegal, lvalue casts are not supported
    (int)z = a;
    ^~~~~~ ~
2 errors generated.
tony@venus:.../perl/git$ cat castlvalue.c
int *x;
int z;

void f(int *y) {
    (int *)x = y;
}

void g(int a) {
    (int)z = a;
}
```

--- a/cop.h
+++ b/cop.h
@@ -813,9 +813,9 @@ struct block_loop {
                 ? &GvSV((c)->blk_loop.itervar_u.gv)     \
                 : (SV **)&(c)->blk_loop.itervar_u.gv)
 
-#define CxLABEL(c)	(0 + CopLABEL((c)->blk_oldcop))
-#define CxLABEL_len(c,len)	(0 + CopLABEL_len((c)->blk_oldcop, len))
-#define CxLABEL_len_flags(c,len,flags)	(0 + CopLABEL_len_flags((c)->blk_oldcop, len, flags))
+#define CxLABEL(c)	(CopLABEL((c)->blk_oldcop))
+#define CxLABEL_len(c,len)	(CopLABEL_len((c)->blk_oldcop, len))
+#define CxLABEL_len_flags(c,len,flags)	((const char *)CopLABEL_len_flags((c)->blk_oldcop, len, flags))
 #define CxHASARGS(c)	(((c)->cx_type & CXp_HASARGS) == CXp_HASARGS)
 
 /* CxLVAL(): the lval flags of the call site: the relevant flag bits from

--- a/cv.h
+++ b/cv.h
@@ -38,7 +38,7 @@ See L<perlguts/Autoloading with XSUBs>.
 #  define Nullcv Null(CV*)
 #endif
 
-#define CvSTASH(sv)	(0+((XPVCV*)MUTABLE_PTR(SvANY(sv)))->xcv_stash)
+#define CvSTASH(sv)	(MUTABLE_HV(((XPVCV*)MUTABLE_PTR(SvANY(sv)))->xcv_stash))
 #define CvSTASH_set(cv,st) Perl_cvstash_set(aTHX_ cv, st)
 #define CvSTART(sv)	((XPVCV*)MUTABLE_PTR(SvANY(sv)))->xcv_start_u.xcv_start
 #define CvROOT(sv)	((XPVCV*)MUTABLE_PTR(SvANY(sv)))->xcv_root_u.xcv_root

--- a/gv.h
+++ b/gv.h
@@ -28,7 +28,7 @@ struct gp {
 
 #if defined (DEBUGGING) && defined(PERL_USE_GCC_BRACE_GROUPS) && !defined(__INTEL_COMPILER)
 #  define GvGP(gv)							\
-        (0+(*({GV *const _gvgp = (GV *) (gv);				\
+        ((GP *)(*({GV *const _gvgp = (GV *) (gv);				\
             assert(SvTYPE(_gvgp) == SVt_PVGV || SvTYPE(_gvgp) == SVt_PVLV); \
             assert(isGV_with_GP(_gvgp));				\
             &((_gvgp)->sv_u.svu_gp);})))
@@ -134,7 +134,7 @@ Return the CV from the GV.
                          GvGP(gv)->gp_hv : \
                          GvGP(gv_HVadd(gv))->gp_hv)
 
-#define GvCV(gv)	(0+GvGP(gv)->gp_cv)
+#define GvCV(gv)	((CV*)GvGP(gv)->gp_cv)
 #define GvCV_set(gv,cv)	(GvGP(gv)->gp_cv = (cv))
 #define GvCVGEN(gv)	(GvGP(gv)->gp_cvgen)
 #define GvCVu(gv)	(GvGP(gv)->gp_cvgen ? NULL : GvGP(gv)->gp_cv)

--- a/peep.c
+++ b/peep.c
@@ -1925,10 +1925,10 @@ S_maybe_multideref(pTHX_ OP *start, OP *orig_o, UV orig_action, U8 hints)
         bool is_last         = FALSE; /* no more derefs to follow */
         bool maybe_aelemfast = FALSE; /* we can replace with aelemfast? */
         UV action_word       = 0;     /* all actions so far */
-        UNOP_AUX_item *arg     = arg_buf;
+        size_t argi          = 0;
         UNOP_AUX_item *action_ptr = arg_buf;
 
-        arg++; /* reserve slot for first action word */
+        argi++; /* reserve slot for first action word */
 
         switch (action) {
         case MDEREF_HV_gvsv_vivify_rv2hv_helem:
@@ -1939,15 +1939,15 @@ S_maybe_multideref(pTHX_ OP *start, OP *orig_o, UV orig_action, U8 hints)
         case MDEREF_AV_gvav_aelem:
             if (pass) {
 #ifdef USE_ITHREADS
-                arg->pad_offset = cPADOPx(start)->op_padix;
+                arg_buf[argi].pad_offset = cPADOPx(start)->op_padix;
                 /* stop it being swiped when nulled */
                 cPADOPx(start)->op_padix = 0;
 #else
-                arg->sv = cSVOPx(start)->op_sv;
+                arg_buf[argi].sv = cSVOPx(start)->op_sv;
                 cSVOPx(start)->op_sv = NULL;
 #endif
             }
-            arg++;
+            argi++;
             break;
 
         case MDEREF_HV_padhv_helem:
@@ -1957,12 +1957,12 @@ S_maybe_multideref(pTHX_ OP *start, OP *orig_o, UV orig_action, U8 hints)
         case MDEREF_AV_padav_aelem:
         case MDEREF_AV_padsv_vivify_rv2av_aelem:
             if (pass) {
-                arg->pad_offset = start->op_targ;
+                arg_buf[argi].pad_offset = start->op_targ;
                 /* we skip setting op_targ = 0 for now, since the intact
                  * OP_PADXV is needed by check_hash_fields_and_hekify */
                 reset_start_targ = TRUE;
             }
-            arg++;
+            argi++;
             break;
 
         case MDEREF_HV_pop_rv2hv_helem:
@@ -2047,8 +2047,8 @@ S_maybe_multideref(pTHX_ OP *start, OP *orig_o, UV orig_action, U8 hints)
                         && o->op_private == 0)
                     {
                         if (pass)
-                            arg->pad_offset = o->op_targ;
-                        arg++;
+                            arg_buf[argi].pad_offset = o->op_targ;
+                        argi++;
                         index_type = MDEREF_INDEX_padsv;
                         o = o->op_next;
                     }
@@ -2087,10 +2087,10 @@ S_maybe_multideref(pTHX_ OP *start, OP *orig_o, UV orig_action, U8 hints)
 #ifdef USE_ITHREADS
                             /* Relocate sv to the pad for thread safety */
                             op_relocate_sv(&cSVOPo->op_sv, &o->op_targ);
-                            arg->pad_offset = o->op_targ;
+                            arg_buf[argi].pad_offset = o->op_targ;
                             o->op_targ = 0;
 #else
-                            arg->sv = cSVOPx_sv(o);
+                            arg_buf[argi].sv = cSVOPx_sv(o);
 #endif
                         }
                     }
@@ -2111,14 +2111,14 @@ S_maybe_multideref(pTHX_ OP *start, OP *orig_o, UV orig_action, U8 hints)
                             maybe_aelemfast = TRUE;
 
                         if (pass) {
-                            arg->iv = iv;
+                            arg_buf[argi].iv = iv;
                             SvREFCNT_dec_NN(cSVOPo->op_sv);
                         }
                     }
                     if (pass)
                         /* we've taken ownership of the SV */
                         cSVOPo->op_sv = NULL;
-                    arg++;
+                    argi++;
                     index_type = MDEREF_INDEX_const;
                     o = o->op_next;
                     break;
@@ -2152,15 +2152,15 @@ S_maybe_multideref(pTHX_ OP *start, OP *orig_o, UV orig_action, U8 hints)
 
                     if (pass) {
 #ifdef USE_ITHREADS
-                        arg->pad_offset = cPADOPx(o)->op_padix;
+                        arg_buf[argi].pad_offset = cPADOPx(o)->op_padix;
                         /* stop it being swiped when nulled */
                         cPADOPx(o)->op_padix = 0;
 #else
-                        arg->sv = cSVOPx(o)->op_sv;
+                        arg_buf[argi].sv = cSVOPx(o)->op_sv;
                         cSVOPo->op_sv = NULL;
 #endif
                     }
-                    arg++;
+                    argi++;
                     index_type = MDEREF_INDEX_gvsv;
                     o = kid->op_next;
                     break;
@@ -2315,7 +2315,7 @@ S_maybe_multideref(pTHX_ OP *start, OP *orig_o, UV orig_action, U8 hints)
                 index_skip = action_count;
                 action |= MDEREF_FLAG_last;
                 if (index_type != MDEREF_INDEX_none)
-                    arg--;
+                    argi--;
             }
 
             action_word |= (action << (action_ix * MDEREF_SHIFT));
@@ -2324,11 +2324,12 @@ S_maybe_multideref(pTHX_ OP *start, OP *orig_o, UV orig_action, U8 hints)
             /* if there's no space for the next action, reserve a new slot
              * for it *before* we start adding args for that action */
             if ((action_ix + 1) * MDEREF_SHIFT > UVSIZE*8) {
-                if (pass)
+                if (pass) {
                     action_ptr->uv = action_word;
+                    action_ptr = arg_buf + argi;
+                }
                 action_word = 0;
-                action_ptr = arg;
-                arg++;
+                argi++;
                 action_ix = 0;
             }
         } /* while !is_last */
@@ -2337,7 +2338,7 @@ S_maybe_multideref(pTHX_ OP *start, OP *orig_o, UV orig_action, U8 hints)
 
         if (!action_ix)
             /* slot reserved for next action word not now needed */
-            arg--;
+            argi--;
         else if (pass)
             action_ptr->uv = action_word;
 
@@ -2516,7 +2517,7 @@ S_maybe_multideref(pTHX_ OP *start, OP *orig_o, UV orig_action, U8 hints)
             op_null(top_op);
         }
         else {
-            Size_t size = arg - arg_buf;
+            Size_t size = argi;
 
             if (maybe_aelemfast && action_count == 1)
                 return;

--- a/pp.c
+++ b/pp.c
@@ -2676,7 +2676,7 @@ PP(pp_i_multiply)
     tryAMAGICbin_MG(mult_amg, AMGf_assign);
     {
       dPOPTOPiirl_nomg;
-      SETi( left * right );
+      SETi( (IV)((UV)left * (UV)right) );
       RETURN;
     }
 }
@@ -2695,7 +2695,7 @@ PP(pp_i_divide)
 
       /* avoid FPE_INTOVF on some platforms when num is IV_MIN */
       if (value == -1)
-          value = - num;
+          value = (IV)-(UV)num;
       else
           value = num / value;
       SETi(value);
@@ -2726,7 +2726,7 @@ PP(pp_i_add)
     tryAMAGICbin_MG(add_amg, AMGf_assign);
     {
       dPOPTOPiirl_ul_nomg;
-      SETi( left + right );
+      SETi( (IV)((UV)left + (UV)right) );
       RETURN;
     }
 }
@@ -2737,7 +2737,7 @@ PP(pp_i_subtract)
     tryAMAGICbin_MG(subtr_amg, AMGf_assign);
     {
       dPOPTOPiirl_ul_nomg;
-      SETi( left - right );
+      SETi( (IV)((UV)left - (UV)right) );
       RETURN;
     }
 }
@@ -2835,7 +2835,7 @@ PP(pp_i_negate)
     {
         SV * const sv = TOPs;
         IV const i = SvIV_nomg(sv);
-        SETi(-i);
+        SETi((IV)-(UV)i);
         return NORMAL;
     }
 }

--- a/sv.h
+++ b/sv.h
@@ -1398,8 +1398,8 @@ object type. Exposed to perl code via Internals::SvREADONLY().
 #ifndef PERL_POISON
 /* Given that these two are new, there can't be any existing code using them
  *  as LVALUEs, so prevent that from happening  */
-#  define SvPVX_mutable(sv)	(0 + (sv)->sv_u.svu_pv)
-#  define SvPVX_const(sv)	((const char*)(0 + (sv)->sv_u.svu_pv))
+#  define SvPVX_mutable(sv)	((char *)((sv)->sv_u.svu_pv))
+#  define SvPVX_const(sv)	((const char*)((sv)->sv_u.svu_pv))
 #else
 /* Except for the poison code, which uses & to scribble over the pointer after
    free() is called.  */

